deps: update openssl to 1.1.1e

hassaanp opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Fixes: #32210

sam-github commented:
Thanks! This needs a few things:

1. Can you rebase this against master? It seems to be branched from a week ago.
2. Can you please follow the directions in doc/guides/maintaining-openssl.md, specifically with respect to the commit messages?
3. When you do the above, you'll see that the commits with contents that are large but essentially machine generated have clear commit messages describing how they were made, and don't mix hand-written and auto-generated content. This makes it possible to review and approve them without reading the contents, because its clear how the content was created. Please split your updates to the configure scripts into a standalone commit, it should be after the sources update, and before the output of the configure process.

hassaanp commented:
> 1. Can you rebase this against master? It seems to be branched from a week ago.
Rebased, and pushed.

> 2. Can you please follow the directions in doc/guides/maintaining-openssl.md, specifically with respect to the commit messages?
Done. The original second commit where I was modifying the scripts has been split and commit messages have been appropriately rewritten according to the guide.

> 3. When you do the above, you'll see that the commits with contents that are large but essentially machine generated have clear commit messages describing how they were made, and don't mix hand-written and auto-generated content. This makes it possible to review and approve them without reading the contents, because its clear how the content was created. Please split your updates to the configure scripts into a standalone commit, it should be after the sources update, and before the output of the configure process.

I have split my configuration updates to a separate commit: [deps: revised paths in scripts to reflect openssl version 1.1.1e](https://github.com/nodejs/node/pull/32328/commits/115128e85140b051254f86a9ad6e29d255ed258f)

sam-github commented:
You didn't rebase, you merged master into your local branch. Next time, do `git rebase upstream/master && git push --force-with-lease` (assuming you are using recommended setup of calling the nodejs/node remote `upstream`).

I would clean this up for you and force push, but it looks like you deselected that permission for me.

If you look at the top of this PR, you'll see it has 47 commits in it! That's not right. Because of the multiple merging from master into your branch, you'd best do an interactive rebase:

> git fetch --all --prune
> git rebase -i  upstream/master

In the interactive, delete all the lines that are commits that _you did not write_. There should be only three. Also, reorder them. After, the commits will look like:

```

core/tls (deps/update-openssl-to-1.1.1e u+5-47) % git log upstream/master..HEAD
commit ea0429b510d1028584b6cf2da0b8e4a4067e7ac1 (HEAD -> deps/update-openssl-to-1.1.1e, pr/32328)
Author: Hassaan Pasha <hassaan.pasha@teamo.io>
Date:   Wed Mar 18 10:09:37 2020 +0500

    deps: update archs files for OpenSSL-1.1.1e
    
    After an OpenSSL source update, all the config files need to be regenerated and
     comitted by:
        $ cd deps/openssl/config
        $ make
        $ git add deps/openssl/config/archs
        $ git add deps/openssl/include

commit 9c4ff6bc9179b7d938c8441b18c901710cfdb490
Author: Hassaan Pasha <hassaan.pasha@teamo.io>
Date:   Wed Mar 18 10:17:22 2020 +0500

    deps: revised paths in scripts to reflect openssl version 1.1.1e
    
    the scripts used by make were modified to correctly reference
    the source files that were originially in crypto/include/internal
    these files are not residing in include/crypto. The base path has
    been left unaltered since that would require too many changes
    
    comitted by:
        $ git add deps/config/Makefile*
        $ git add deps/config/generate_gypi.pl
        $ git add deps/openssl/openssl/Makefile
        $ git add deps/openssl/openssl/configdata.pm
        $ git commit

commit 4e0ad52ddec2bda9b2144fcb89ec4d6e0a9732ab
Author: Hassaan Pasha <hassaan.pasha@teamo.io>
Date:   Tue Mar 17 22:14:56 2020 +0500

    deps: upgrade openssl sources to 1.1.1e
```

Note that last commit doesn't match the format in the guide: https://github.com/nodejs/node/blob/master/doc/guides/maintaining-openssl.md#1-obtain-and-extract-new-openssl-sources

We aren't normally this nitpicky, but the openssl maintenance should match the guide, feel free to kick this back to me and/or give me write permissions on your branch if you'd like me to do the rebasing.

ExE-Boss commented:
> You didn't rebase, you merged master into your local branch.

If that was what was done, then there would only be a merge commit.

What instead happened was that `master` was rebased on top of this PR.

hassaanp commented:
@sam acknowledged.
I will follow your suggestions.
I am also going to give you write access incase I am unable to sort this.
Thank you for your patience.

sam-github commented:
Cleaned up and repushed.

sam-github commented:
OK, needs one more commit to (at least) fix test.sequential/test-tls-psk-client?

https://ci.nodejs.org/job/node-test-commit-linux-containered/nodes=ubuntu1804_sharedlibs_zlib_x64/18781/

hassaanp commented:
@sam-github thanks for the help
Looking at the test now.

hassaanp commented:
@sam-github, I have fixed the test but the behavior displayed is a little awkward.
The socket `end` event is not fired until the `data` event is consumed. I hope I am explaining this correctly. Pushing the code, hopefully that would make more sense.

hassaanp commented:
I have pushed the code. Kindly review.

kapouer commented:
Building 10.19.0 against shared openssl 1.1.1d i got an odd situation because tests ran against openssl 1.1.1e, and only one js test is failing: `parallel/test-tls-session-cache` (ppc64el but i'm not sure it's a problem with the arch in itself):
https://buildd.debian.org/status/fetch.php?pkg=nodejs&arch=ppc64el&ver=10.19.0~dfsg-2&stamp=1584648000&raw=0

MylesBorins commented:
Still seeing lots of failures in CI on this. Would be very good for us to include this in the various releases that we have targeting next tuesday. Do folks need help with moving this forward so we can meet that timeline? We likely need backport PRs to test each release line

sam-github commented:
Pushed a fix for test/parallel/test-tls-session-cache.js

jasnell commented on test/sequential/test-tls-psk-client.js:
```diff
@@ -37,6 +37,7 @@ const timeout = setTimeout(() => cleanUp('Timeouted'), 5000);
 
 function waitForPort(port, cb) {
   const socket = net.connect(common.PORT, () => {
+    socket.on('data', ()=>{})
```
```suggestion
    socket.on('data', () => {})
```

kapouer commented:
> Pushed a fix for test/parallel/test-tls-session-cache.js

Works on debian/sid builds.

sam-github commented:
@hassaanp  Our PR landing tooling isn't happy, I think its because the commits are authored by "Hassaan Pasha <hassaan.pasha@teamo.io>", so may I change the email to the address on your github page, https://github.com/hassaanp? Alternatively, can you check your github account setup to ensure that github knows that `hassaan.pasha@teamo.io` is confirmed as one of your addresses? See https://github.com/settings/emails

jasnell commented:
The QUIC modifications patch required only a couple of minor modifications to successfully build with this, which is fantastic. I'm +1 on this landing when it's ready

hassaanp commented:
@sam-github, sorry about that. I have added the email address to my Github.
Please confirm if this checks out.

sam-github commented:
@hassaanp that worked great, git node land is happy, this could land if it had some reviews.

@MylesBorins @jasnell 

> This PR needs to wait 93 more hours to land (or 0 hours if there is one more approval)

kapouer commented:
@sam-github i just uploaded 10.19.0 with shared openssl 1.1.1e and the test fix.
It's not exactly like your ci but... if all builds pass it's usually a very good sign ;)
https://buildd.debian.org/status/package.php?p=nodejs

kapouer commented:
...(lockdown gives me time for opensource again)....

jasnell commented:
In the process of landing this now...

sam-github commented:
git node land says its not ready, but it doesn't say why:

```
core/tls (master % u=) % git node land https://github.com/nodejs/node/pull/32328
✔  Done loading data for nodejs/node/pull/32328
----------------------------------- PR info ------------------------------------
Title      deps: update openssl to 1.1.1e (#32328)
Author     Hassaan Pasha <pasha.hassaan@gmail.com> (@hassaanp, first-time contributor)
Branch     hassaanp:deps/update-openssl-to-1.1.1e -> nodejs:master
Labels     openssl
Commits    5
 - deps: upgrade openssl sources to 1.1.1e
 - deps: adjust openssl configuration for 1.1.1e
 - deps: update archs files for OpenSSL-1.1.1e
 - test: discard data received by client
 - test: end tls connection with some data
Committers 1
 - Sam Roberts <vieuxtech@gmail.com>
------------------------------ Generated metadata ------------------------------
PR-URL: https://github.com/nodejs/node/pull/32328
Fixes: https://github.com/nodejs/node/issues/32210
Reviewed-By: Sam Roberts <vieuxtech@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
--------------------------------------------------------------------------------
   ℹ  Last Full PR CI on 2020-03-20T14:24:56Z: https://ci.nodejs.org/job/node-test-pull-request/29951/
   ℹ  This PR was created on Tue, 17 Mar 2020 17:41:51 GMT
   ✔  Approvals: 2
   ✔  - Sam Roberts (@sam-github) (TSC): https://github.com/nodejs/node/pull/32328#pullrequestreview-378698239
   ✔  - James M Snell (@jasnell) (TSC): https://github.com/nodejs/node/pull/32328#pullrequestreview-379485001
   ⚠  PR author is a new contributor: @hassaanp(pasha.hassaan@gmail.com)
   ⚠  - commit 9a8170900927 is authored by hassaan.pasha@teamo.io
   ⚠  - commit 02041c786d85 is authored by hassaan.pasha@teamo.io
   ⚠  - commit 2ac8f0dc4d79 is authored by hassaan.pasha@teamo.io
   ⚠  - commit b7ca3d39bf64 is authored by hassaan.pasha@teamo.io
--------------------------------------------------------------------------------
? This PR is not ready to land, do you want to continue? (Y/n) 
```

Having a PR by a new contributor isn't blocking, so what's its beef with the PR?

jasnell commented:
Saw that as well, as far as I can determine it's just the email account, which I don't see as being critical.

addaleax commented:
@sam-github Maybe it’s complaining about the mismatch between the git `Author` fields and the public github email? I think it should be okay to ignore that.

jasnell commented:
Landed in 116a30b, 425b962, ec144d5, 3d0f225, and 6b5f82d

sam-github commented:
@jasnell Thanks.

@addaleax It used to complain differently about the emails, but @hassaanp changed something, and  that went away. Hm.

jasnell commented:
@sam-github, yeah, not really sure but github has been just a bit flaky this morning for me so I'm wondering if there's some error happening there that's not surfacing quite correctly in the tool

sam-github commented:
For future historians: note that 6b5f82d7f8 turns out to have been necessary to work around a recently introduced behaviour change in s_client, see discussion in [openssl issue 11378](https://github.com/openssl/openssl/issues/11378)
